sudoku-django-project/mainapp/classes.py

```diff
@@ -0,0 +1,283 @@
+from random import shuffle, randint
+import typing
+
+BoardType = typing.List[typing.List[typing.Union[None, int]]]    # type hint alias for board - a list of lists
+                                                                 # with int or None as values
+class Sudoku:
+    """" A class that acts like a sudoku puzzle. """
+
+    def __init__(self):  # noqa: ANN204
+        self.counter = 1
+        self.top_boxes = [_Box() for _ in range(3)]
+        self.mid_boxes = [_Box() for _ in range(3)]
+        self.bottom_boxes = [_Box() for _ in range(3)]
+
+        self._set_box_locations()
+
+        self.original = [[1, 2, 3, 4, 5, 6, 7, 8, 9],
```

Comment:
Can be made smaller with list comprehension

Reply:
Maybe make it a separate function, so the object's attribute only contains the actual puzzle and not the thing you start off with

Reply:
Yeah having an attribute for the original seems unnecessary. How would a list comp work though? The first inner list is simple enough, but the rest are shuffled :thinking:

Reply:
I'm reluctant to use list comp just for one of the inner elements and the hardcode the rest, feel like it'll look...ugly.

Reply:
all the inner lists are 1-9 right, so you could do
`[[i for i in range(1,10)] for i in range(1,10)]`

Reply:
this is supposed to be a working board, i.e., should be able to be used as a puzzle itself, hence it is hard coded.

Reply:
The order in the rest is different, but the shuffle isn't random, so we can't `random.shuffle` it (I think, I'm not very sure how this works)

Reply:
So the actual puzzle after generation is stored in the original itself?

Reply:
Unsure, but if so that's now a misleading variable name

Reply:
Also what does the fullboard attribute hold then?

Reply:
The way the generator works is that the class object holds the shuffled version (in a way that creates a board) of the hard coded board, and it was used for shorter computation times (compared to back tracking for generation). The original board wont change value, but the instance of the class (self) itself will store the shuffled board.

Reply:
so what do the original board and fullboard attributes hold exactly?

Reply:
@Windsmith Original stores the original board that is used to generate a separate board, full_board stores the newly generated board with all the solutions, and I guess puzzle_board is not really necessary, so I can remove that variable

Reply:
puzzle_board variable?
Also I think its better if you turn the original board into a separate function that just returns that board. I doubt its good to leave multiple board configurations as attributes for a class supposed to represent a single board.
You could make it a static method.